veps/sig-compute/10-dra-devices/vep.md

```diff
@@ -0,0 +1,781 @@
+## Design
+
+For allowing users to consume DRA devices, there are two main changes needed:
+
+1. API changes and the plumbing required in KubeVirt to generate the domain xml with the devices.
+2. Driver Implementation to set the required attributes for KubeVirt to use
+
+This design document focuses on part 1 of the problem. We're introducing a new feature gate `DRADevices`.
+All the API changes will be gated behind this feature gate so as not to break existing functionality.
+
+### API Changes
+
+```go
+type VirtualMachineInstanceSpec struct {
+	..
+	..
+	// ResourceClaims defines which ResourceClaims must be allocated
+	// and reserved before the VMI and hence virt-launcher pod is allowed to start. The resources
+	// will be made available to the domain which consume them
+	// by name.
+	//
+	// This is an alpha field and requires enabling the
+	// DynamicResourceAllocation feature gate in kubernetes
+	//  https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/
+	//
+	// This field is immutable.
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	ResourceClaims []k8sv1.PodResourceClaim `json:"resourceClaims,omitempty"`
+}
+
+type GPU struct {
+	// Name of the GPU device as exposed by a device plugin
+	Name string `json:"name"`
+	// DeviceSource is the name of the device provisioned either by device plugins
+	// or by DRA enabled device
+	// DeviceName string `json:"deviceName"`    <-- inlined into DeviceSource
+	DeviceSource      DeviceSource `json:",inline"`
+	VirtualGPUOptions *VGPUOptions `json:"virtualGPUOptions,omitempty"`
+	// If specified, the virtual network interface address and its tag will be provided to the guest via config drive
+	// +optional
+	Tag string `json:"tag,omitempty"`
+}
+
+type HostDevice struct {
+	Name string               `json:"name"`
+	// DeviceSource is the name of the device provisioned either by device plugins
+	// or by DRA enabled device
+	// DeviceName string `json:"deviceName"`    <-- inlined into DeviceSource
+	DeviceSource DeviceSource `json:",inline"`
+	// If specified, the virtual network interface address and its tag will be provided to the guest via config drive
+	// +optional
+	Tag string `json:"tag,omitempty"`
+}
+
+type DeviceSource struct {
+	// DeviceName is the name of the device provisioned by device-plugins
+	DeviceName *string `json:"deviceName,omitempty"`
+	// ClaimRequest provides the ClaimName from vmi.spec.resourceClaims[].name and
+	// requestName from resourceClaim.spec.devices.requests[].name
+	// this fields requires DRA feature gate enabled
+	ClaimRequest *ClaimRequest `json:",inline,omitempty"`
+}
+
+type ClaimRequest struct {
+	// ClaimName needs to be provided from the list vmi.spec.resourceClaims[].name where this
+	// device is allocated
+	ClaimName string `json:"claimName"`
+	// RequestName needs to be provided from resourceClaim.spec.devices.requests[].name where this
+	// device is requested
+	RequestName string `json:"requestName"`
+}
```

Comment:
Do we have the API that was discussed on the last meeting?

Reply:
This was the agreed upon API from the meeting.